### PR TITLE
Add a redirect history

### PR DESCRIPTION
**Pull request by dawehner**
See https://github.com/guzzle/guzzle/issues/1166#issuecomment-122248990

### CODE REVIEW

**Comment by dawehner** [comment]
Mh, this seems to be unrelated to that patch

File: tests/RedirectMiddlewareTest.php [hunk 1]
```diff
@@ -128,6 +129,41 @@ public function testAddsRefererHeader()
...
+            'allow_redirects' => [
+                  'track_redirects' => true,
+            ],
+        ]);
+        /** @var ResponseInterface $response */
+        $response = $promise->wait(TRUE);
+
+        $this->assertEquals(
+          'http://example.com?a=b',
+          $response->getHeaderLine('X-Guzzle-Redirect-0')
```
Comment (Crell) [inline tests/RedirectMiddlewareTest.php]: Do we want the headers distinctly named or should they share a name?  Distinctly named is easier if you know what you're looking for.  Shared name means they come back as a collection, which is better if you don't know exactly how many you're working with.

I think I'd marginally prefer a shared header, but defer to @mtdowling

Comment (mtdowling) [inline tests/RedirectMiddlewareTest.php]: I agree. I'd prefer a shared header.

**Comment by mtdowling** [comment]
What do you think, @dawehner? Can we use the same header and just append a value to the header? Using the same header field to track ordered values should be acceptable based on RFC 7230:

```
  A recipient MAY combine multiple header fields with the same field
   name into one "field-name: field-value" pair, without changing the
   semantics of the message, by appending each subsequent field value to
   the combined field value in order, separated by a comma.  The order
   in which header fields with the same field name are received is
   therefore significant to the interpretation of the combined field
   value; a proxy MUST NOT change the order of these field values when
   forwarding a message.
```

See http://tools.ietf.org/html/rfc7230#section-3.2.2

**Comment by dawehner** [comment]
Does that look more like you expect it to look like?

File: src/RedirectMiddleware.php [hunk 2]
```diff
@@ -98,7 +99,18 @@ public function checkRedirect(
...
         }
 
-        return $this($nextRequest, $options);
+        /** @var PromiseInterface|ResponseInterface $response */
+        $options['_previous_uris'][] = $request->getUri();
+        $response = $this($nextRequest, $options);
+
+        // Add headers to be able to track history of redirects.
+        if (!empty($options['allow_redirects']['track_redirects'])) {
+            $response = $response->wait(TRUE);
```
Comment (mtdowling) [inline src/RedirectMiddleware.php]: We don't want to wait here. Instead, you should return a new promise that modifies the response before it is returned.

File: src/RedirectMiddleware.php [hunk 3]
```diff
@@ -98,7 +99,18 @@ public function checkRedirect(
...
+            if (!$response->hasHeader('X-Guzzle-Redirect')) {
```
Comment (mtdowling) [inline src/RedirectMiddleware.php]: Instead of this, just use the `withAddedHeader` method of a response and remove the if statement.

**Comment by mtdowling** [comment]
Any news here?

**Comment by mtdowling** [comment]
Any update?

**Comment by dawehner** [comment]
Not yet sorry. I need to fine some time, but in case anyone wants to, feel
free to take it over.

File: src/RedirectMiddleware.php [hunk 4]
```diff
@@ -21,7 +21,8 @@ class RedirectMiddleware
         'max'       => 5,
         'protocols' => ['http', 'https'],
         'strict'    => false,
-        'referer'   => false
+        'referer'   => false,
+        'track_redirects' => false,
```
Comment (GrahamCampbell) [inline src/RedirectMiddleware.php]: odd how the rest of the array is aligned, but this isn't

File: tests/RedirectMiddlewareTest.php [hunk 5]
```diff
@@ -128,6 +129,30 @@ public function testAddsRefererHeader()
         );
     }
 
+    public function testAddsGuzzleRedirectHeader()
+    {
+        $mock = new MockHandler([
+          new Response(302, ['Location' => 'http://example.com']),
```
Comment (GrahamCampbell) [inline tests/RedirectMiddlewareTest.php]: you need 4 spaces for indentation, not 2

File: src/RedirectMiddleware.php (hunk 2 above)
Comment (dawehner) [inline src/RedirectMiddleware.php]: I tried really hard to use a promise, but I simply cannot figure it out properly.
The problem is that the most outer promise will need to have the information of all redirects. On the other hand the most outer promise is created within the scope of none already made redirect.

The most inner promise though has the information of all redirects available, but I have absolutely no clue at that point, how to pass down that information. For quite some time I considered the $response from checkRedirect as the previous response, but this is actually not the case. Its, as you know, a response created for manipulation only, something which is designed in a way that the API is independent of the concrete PSR7 implementation.

File: src/RedirectMiddleware.php (hunk 3 above)
Comment (dawehner) [inline src/RedirectMiddleware.php]: Well, this doesn't work as you expect it. With this if() we ensure that just the most outer response, so the one all the previous URIs available, sets the header, I think, but swapping just that lets the test fail

**Comment by stefanruijsenaars** [comment]
Added a PR that includes a fix for the conflict and the coding standard fixes at https://github.com/guzzle/guzzle/issues/1244

**Comment by mtdowling** [comment]
I cleaned this up and made a few tweaks.

**Comment by dawehner** [comment]
You are truly a great opensource maintainer! The project queue is cleaned up regularly, you respond quickly, you are friendly etc. Thank you for being awesome

**Comment by mtdowling** [comment]
Thanks, @dawehner :)

**Pull request merged**